src/lib.rs

```diff
@@ -454,7 +489,21 @@ impl<'object, Endian> DebugInfo<'object, Endian>
                 // matches us better. We need to keep going until we find the best one.
             }
 
-            return Ok(Some((path, line, func.map(|u| u.0.name.to_string_lossy()))));
+            let func = func.map(|u| {
+                if self.with_demangling {
+                    // First try demangling as a C++ symbol, then fall back to
+                    // Rust symbols.
+                    if let Ok(sym) = cpp_demangle::Symbol::new(u.0.name.to_bytes()) {
+                        Cow::from(format!("{}", sym))
+                    } else {
+                        Cow::from(format!("{}", demangle(u.0.name.to_string_lossy().as_ref())))
+                    }
+                } else {
+                    u.0.name.to_string_lossy()
+                }
+            });
```

Comment:
So, I've been going back and forth on this now a bunch (you may have seen some comments appear and then disappear). I really dislike that demangling is pushed this far into `addr2line`, when it can be almost entirely externally. Admittedly, if we want to use `DW_AT_language`, we may have to do it here, but if not, this could be done in `Mapping::locate`. At the very least I'd like to see the closure above moved into a function.